### PR TITLE
TupleExpression implementation

### PR DESCRIPTION
**Pull request by Praveen2112**
This is the implementation that we had done for supporting OR predicate in TupleDomain. If this implementation is proper then we can modify the DomainTranslator.java and also at each connector side.

### CODE REVIEW

**Comment by KBP-TDC** [comment]
@Praveen2112 Would you mind to describe what kind of expressions this PR would enable (a few examples  of various complexity would be best!) and also describe the limitations (if any, with examples as well).
I also see no tests. Is there anything you would wish to introduce to ensure correctness?

**Comment by maciejgrzybek** [comment]
Someone from Neutrino (@cawallin, @arhimondr, @amrutagokhale or me) will have a look on that.

**Comment by Praveen2112** [comment]
@KBP-TDC This PR would enable the expressions which are handled on TupleDomain along with OR conditions.
Eg. a = 1 || b = 2.

Ideally the Expressions which are handled by TupleDomain can be handled in the TupleExpression.

The expressions which cannot be handled are 
1. Filter criteria applied over the results of a function 
Eg. function(a) > 23
a + b =10
These can be handled but they are not supported currently.

**Comment by Praveen2112** [comment]
@KBP-TDC We have implemented the TupleExpression for Hive Connector (Orc files). For that we have tweaked the TupleDomain to hold TupleExpression instead of the Map<ColumnHandle,Domain>. It also involves changing the code at the optimizer level and modifying `TupleDomainOrcPredicate.java` so that we would read the stripes that satisfy the filter criteria.

**Comment by KBP-TDC** [comment]
@Praveen2112 Thank you for the explanation. 
What I am most curious about is how complex the predicate can be in your implementation. 
Say, would both of the following:
1) col_1 = 5 OR (col_2 > 3 AND col_2 < 10) OR col_3 BETWEEN 7 AND 15
2) (col_a = 2 OR col_a = 6) AND (col_b = 4 OR col_b = 11) AND col_c IN (15, 19, 23)
work fine (i.e., be pushed down to ORC reader)?

**Comment by Praveen2112** [comment]
@KBP-TDC Yes. Both can be implemented. The expression 2 is currently implemented in TupleDomain. Using TupleExpression we can implement both.

File: presto-spi/src/main/java/com/facebook/presto/spi/predicate/OrExpression.java [hunk 1]
```diff
@@ -80,6 +80,18 @@ public TupleExpression transform(Function function)
...
     @Override
+    public boolean isAll()
+    {
+        return leftExpression.isAll() || rightExpression.isAll();
+    }
+
+    @Override
+    public boolean isNone()
+    {
+        return leftExpression.isNone() && rightExpression.isNone();
```
Comment (Praveen2112) [inline presto-spi/src/main/java/com/facebook/presto/spi/predicate/OrExpression.java]: We can assume an OrExpression to be should not have any values(`isNone()` represents that) only if both the operands have no values in their domain. So we cant use `||` here.

File: presto-spi/src/main/java/com/facebook/presto/spi/predicate/AndExpression.java [hunk 2]
```diff
@@ -1,8 +1,108 @@
...
+        Map<C, NullableValue> map = new HashMap<>();
+        map.putAll(leftExpression.extractFixedValues());
+        map.putAll(rightExpression.extractFixedValues());
+        return map;
+    }
+
+    @Override
+    public List<C> getColumnDomains()
+    {
+        List<C> columnList = new ArrayList<C>();
```
Comment (Praveen2112) [inline presto-spi/src/main/java/com/facebook/presto/spi/predicate/AndExpression.java]: Well for Presto-spi module we should have minimal dependencies...so used it....If we need to use` ImmutableList.Builder` then I think we should add guava as dependency. As of guava library is used only for testing spi.

File: presto-spi/src/main/java/com/facebook/presto/spi/predicate/AndExpression.java [hunk 3]
```diff
@@ -1,8 +1,108 @@
...
+    @JsonProperty
+    public TupleExpression<C> getRightExpression()
+    {
+        return rightExpression;
+    }
+
+    @Override
+    public Map<C, NullableValue> extractFixedValues()
+    {
+        Map<C, NullableValue> map = new HashMap<>();
```
Comment (Praveen2112) [inline presto-spi/src/main/java/com/facebook/presto/spi/predicate/AndExpression.java]: Well for Presto-spi module we should have minimal dependencies...so used it....If we need to use` ImmutableMap.Builder` then I think we should add guava as dependency. As of now guava library is used only for testing spi.

File: presto-spi/src/main/java/com/facebook/presto/spi/predicate/AndExpression.java [hunk 4]
```diff
@@ -87,6 +87,18 @@ public String getName()
     }
 
     @Override
+    public boolean isNone()
+    {
+        return leftExpression.isNone() || rightExpression.isNone();
```
Comment (Praveen2112) [inline presto-spi/src/main/java/com/facebook/presto/spi/predicate/AndExpression.java]: We can assume an `AndExpression` to be should not have any values(isNone() represents that) if any of the operands have no values in their domain. So we cant use && here.

File: presto-spi/src/main/java/com/facebook/presto/spi/predicate/DomainExpression.java [hunk 5]
```diff
@@ -1,8 +1,112 @@
...
+    @JsonProperty
+    public Domain getDomain()
+    {
+        return domain;
+    }
+
+    @Override
+    public Map<C, NullableValue> extractFixedValues()
+    {
+        Map<C, NullableValue> map = new HashMap();
```
Comment (Praveen2112) [inline presto-spi/src/main/java/com/facebook/presto/spi/predicate/DomainExpression.java]: Well for Presto-spi module we should have minimal dependencies...so used it....If we need to use `ImmutableMap.Builder `then I think we should add guava as dependency. As of now guava library is used only for testing spi.

**Comment by KBP-TDC** [comment]
Praveen2112 How about Parquet? Especially the new optimized reader. Will your PR get a similar benefits for Parquet as it does for ORC?

**Comment by Praveen2112** [comment]
@KBP-TDC Yes..Ideally for each connector and each file format we have to implement `TupleExpressionVisitor` then we will get the similar benefits as it does for ORC.

File: presto-spi/src/main/java/com/facebook/presto/spi/predicate/AndExpression.java (hunk 3 above)
Comment (maciejgrzybek) [inline presto-spi/src/main/java/com/facebook/presto/spi/predicate/AndExpression.java]: Fair, I overlooked that. Don't introduce that dependency to the SPI.

**Comment by Praveen2112** [comment]
@maciejgrzybek Any other things that we have to change.

File: presto-spi/src/main/java/com/facebook/presto/spi/predicate/NotExpression.java [hunk 6]
```diff
@@ -62,6 +62,16 @@ public TupleExpression transform(Function function)
...
     }
 
+    public boolean isNone()
+    {
+        return expression instanceof AllExpression || expression.isAll();
+    }
+
+    public boolean isAll()
+    {
+        return expression instanceof NoneExpression || expression.isNone();
```
Comment (Praveen2112) [inline presto-spi/src/main/java/com/facebook/presto/spi/predicate/NotExpression.java]: None ensures that no value is selected (No value is selected in a stripe). So Not(None) = all...(All ensures that all values are selected (Select the entire stripes).

**Comment by Praveen2112** [comment]
We use TupleExpression as replacement for TupleDomian. So it involves modifying the code base at a larger level in the Query planning (from the TableScanNode) , in the optimizers and finally at the connector side. We did those changes only for Hive connector and used it. As for the use case we are trying to support OR expression for Predicate push down in Presto where as the TupleDomain (current implementation) supports only AND expression.

**Comment by maciejgrzybek** [comment]
@Praveen2112 why don't you add the other other bits of your code to the PR, namely the changes to the planner (to be able to express predicates in `TableScanNode` as `TupleExpression`) and optimizers doing the pushdown?
Because without it, it'd be a dead code if we merged it.

How big is your deployment? BTW What company do you work for?

**Comment by Praveen2112** [comment]
@maciejgrzybek I ll add the other bits of code. I just want to ensure that is this a right approach to solve this problem before modifying the entire code base. We had modified it only for hive connector(handled only for non partitioned case).I have just started checking out presto for few use cases and I work with Zoho.

**Comment by Praveen2112** [comment]
@maciejgrzybek In the above implementation OrExpression is handled only for Hive connector and that too for non partitioned orc files.

**Comment by Praveen2112** [comment]
@maciejgrzybek I ll close this PR and open it as a new PR

**Comment by maciejgrzybek** [comment]
Thanks, please assign me when it's done.

**Pull request closed without merging**